pallets/settlement/src/lib.rs

```diff
@@ -1047,12 +1141,32 @@ decl_module! {
 }
 
 impl<T: Config> Module<T> {
-    fn lock_via_leg(leg: &Leg) -> DispatchResult {
-        T::Portfolio::lock_tokens(&leg.from, &leg.asset, leg.amount)
+    fn lock_via_leg(leg: &LegV2) -> DispatchResult {
+        match &leg.asset {
+            LegAsset::Fungible { ticker, amount } => {
+                T::Portfolio::lock_tokens(&leg.from, &ticker, *amount)?;
+            }
+            LegAsset::NonFungible(nfts) => {
+                for nft_id in nfts.ids() {
+                    T::Portfolio::lock_nft(&leg.from, nfts.ticker(), &nft_id)?;
+                }
```

Comment:
This could lock some NFT before throwing an error.  Then some NFTs would be locked and others unlocked.

Same for the `unlock_via_leg` below.

Reply:
This would be much easier to handle when all extrinsic calls are wrapped in a transaction, like is done with the newer Substrate.